contrib/babelfishpg_tsql/src/hooks.c

```diff
@@ -5276,56 +5276,120 @@ transform_pivot_clause(ParseState *pstate, SelectStmt *stmt)
+	Assert(is_integer_type(int_type1));
+	Assert(is_integer_type(int_type2));
+
+	switch (int_type1)
+	{
+		case INT2OID:
+			return int_type2 == INT4OID || int_type2 == INT8OID;
+		case INT4OID:
+			return int_type2 == INT8OID;
+		default:
+		  Assert(int_type1 == INT8OID);
+		  return false;
+	}
+}
+
 static Node* optimize_explicit_cast(ParseState *pstate, Node *node)
 {
 	if (sql_dialect != SQL_DIALECT_TSQL)
 		return node;
 	if (node == NULL)
 		return NULL;
-	if (IsA(node, FuncExpr))
+
+	if (IsA(node, FuncExpr) &&
+		((FuncExpr *) node)->funcformat == COERCE_EXPLICIT_CAST)
 	{
-		FuncExpr   *f = (FuncExpr *) node;
-		if (f->funcformat == COERCE_EXPLICIT_CAST){
-			Node* arg = (Node*)linitial(f->args);
-			if (nodeTag(arg) == T_Var && 
-			   ((Var*) arg)->vartype == INT4OID && 
-			   f->funcresulttype == INT8OID)
-				return optimize_explicit_cast(pstate, linitial(f->args));
+		Node *original_node = node;
+		/*
+		 * Record the integer result type with lowest precision during CAST
+		 * process. If the original SQL casts a variable of integer type to a
+		 * type with less precision, removing the CASE funcition will lead to
+		 * different results. Initial value as highest precision integer type.
+		 */
+		Oid lowest_precision_type = INT8OID;
+		/*
+		 * Deal with nested CAST functions.
+		 * For example: CAST((CAST col_int AS BIGINT) AS INT) => col_int
+		 */
+		while (((FuncExpr *) node)->funcformat == COERCE_EXPLICIT_CAST) {
+			FuncExpr *f = (FuncExpr *) node;
+
+			if (!is_integer_type(f->funcresulttype)) break;
+
+			if (lower_precision_than(f->funcresulttype, lowest_precision_type))
+				lowest_precision_type = f->funcresulttype;
+
+			node = (Node *) linitial(f->args);
+
+			if (!IsA(node, FuncExpr)) break;
 		}
+
+		/*
+		 * Casting a integer variable to a type with higher precision is
+		 * unnecessary. Optimizer will figure out the appropriate operator to
+		 * calculate without loss of precision.
+		 */
+		if (IsA(node, Var) && is_integer_type(((Var *) node)->vartype) &&
+			!lower_precision_than(lowest_precision_type,
+								  ((Var *) node)->vartype))
+			return node;
+		else
+			return original_node;
 	}
 	else if (IsA(node, BoolExpr))
 	{
 		BoolExpr *r = (BoolExpr *) node;
 		ListCell *l;
-		foreach (l , r->args)
+		foreach (l, r->args)
 		{
-			optimize_explicit_cast(pstate, (Node*)lfirst(l));
+			Node *new_node = optimize_explicit_cast(pstate, (Node*)lfirst(l));
+			if (lfirst(l) != new_node) lfirst(l) = new_node;
 		}
 	}
 	else if (IsA(node, OpExpr))
 	{
 		OpExpr *opExpr = (OpExpr*) node;
-		Form_pg_operator form;
-		HeapTuple	tuple;
-		Node* node1 = optimize_explicit_cast(pstate, linitial(opExpr->args));
-		Node* result = NULL;
-		if (node1 != linitial(opExpr->args))
+		Node *first_old_arg, *fist_new_arg, *second_old_arg, *second_new_arg;
+
+		/*
+		 * Only the following comparsion operators could make use of index after
+		 * removing unnecessary CAST function, which are our targets.
+		 */
```

Comment:
The reasons why I choose to implement in current way have:
1. This feature is a minor optimization while seldom hand-written SQLs could involve such unnecessary CAST functions. I think most of eligible SQLs are generated by framework. I wanna reduce the risk of applying it for all operators. Though I think it'll be okay, I couldn't have the enough tests to ensure that.
2. Only when the existence of CAST function influences the execution plan causing huge performance difference, the customer will notice that. That's the reason why we have this work. If no execution plan changes, I don't think a CAST function will bring observable performance difference for most cases. So I only solve the scenarios where could influence the index choosing.
3. I guess the syscache lookup cannot be avoided. Because we need to make_op according to the new types of arguments, which relies on op_name.

As we talked offline, support only built-in =/</<=/>/>= operators.